Convert SIL.Core and SIL.WritingSystems to .NET Standard 2.0 libraries

ddaspit opened the pull request:
- Multi-target SIL.Core and SIL.WritingSystems to .NET 4.6.1 and .NET Standard 2.0
- Update libpalaso version to 4.1
- MSBuild 15.3+ and NuGet 4.3+ is now required to build (Mono 5.8+ on Linux)
- Add Pack target to Palaso.proj to generate NuGet packages for SIL.Core and SIL.WritingSystems
- Converted all platform-specific code in SIL.Core to be run-time conditional (platform-specific builds are no longer required for SIL.Core)
- Moved all desktop app-specific code in SIL.Core to SIL.Core.Desktop
    - Split IO classes into general and desktop-specific classes (maintained APIs, but marked some methods as obsolete)
    - Moved event reporting, logging, and usage reporting to SIL.Core.Desktop
- Updated to latest version of Json.NET

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/sillsdev/libpalaso/622)
<!-- Reviewable:end -->

ermshiperete commented:
Reviewed 152 of 188 files at r1.
Review status: 115 of 151 files reviewed at latest revision, 23 unresolved discussions, some commit checks failed.

---

*[NuGet.config, line 4 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/622#-L6LigHm4eQ183VZb1qT:-L6LigHn6pgYgMbz79ZS:b-sf09fp) ([raw file](https://github.com/sillsdev/libpalaso/blob/9eac80d73b6633abd88543988053b9e72221ad93/NuGet.config#L4)):*
> ```XML
> <configuration>
>   <packageSources>
>     <add key="build.palaso.org" value="http://build.palaso.org/guestAuth/app/nuget/v1/FeedService.svc/" />
> ```

It would be better to use https

---

*[build/NuGet.targets, line 14 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/622#-L6Lj8ij6FPezdFsequG:-L6Lj8ij6FPezdFsequH:bdv8858) ([raw file](https://github.com/sillsdev/libpalaso/blob/9eac80d73b6633abd88543988053b9e72221ad93/build/NuGet.targets#L14)):*
> ```XML
> 		<!-- Download NuGet.exe if it does not already exist -->
> 		<DownloadNuGetExe Condition=" '$(DownloadNuGetExe)' == '' ">true</DownloadNuGetExe>
>     <NuGetExeUrl>https://dist.nuget.org/win-x86-commandline/latest/nuget.exe</NuGetExeUrl>
> ```

nit-pick: indent with tabs instead of spaces

---

*[SIL.BuildTasks/StampAssemblies/StampAssemblies.cs, line 151 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/622#-L6LlWk3VMr9aa0lgZcj:-L6LlWk3VMr9aa0lgZck:bb9xvp) ([raw file](https://github.com/sillsdev/libpalaso/blob/9eac80d73b6633abd88543988053b9e72221ad93/SIL.BuildTasks/StampAssemblies/StampAssemblies.cs#L151)):*
> ```Smalltalk
> 			{
> 				Parts = (string[]) existing.Parts.Clone(),
> 				Prerelease = existing.Prerelease
> ```

Could write this as

    Prerelease = incoming.Prerelease ?? existing.Prerelease

if we don't want the `if` below.

---

*[SIL.Core/SIL.Core.csproj, line 1 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/622#-L6LmMVSozFF8K1wju2w:-L6LmMVTpCEGRccbrIIJ:b-30foyx) ([raw file](https://github.com/sillsdev/libpalaso/blob/9eac80d73b6633abd88543988053b9e72221ad93/SIL.Core/SIL.Core.csproj#L1)):*
> ```XML
> <Project Sdk="Microsoft.NET.Sdk">
> ```

So much shorter. Nice! :smile: 

---

*[SIL.Core.Desktop.Tests/IO/DirectoryUtilitiesTests.cs, line 19 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/622#-L6LmlfVqKdzCOpMkplg:-L6LmlfVqKdzCOpMkplh:b7zm9x1) ([raw file](https://github.com/sillsdev/libpalaso/blob/9eac80d73b6633abd88543988053b9e72221ad93/SIL.Core.Desktop.Tests/IO/DirectoryUtilitiesTests.cs#L19)):*
> ```Smalltalk
> 		{
> 			ErrorReport.IsOkToInteractWithUser = false;
> 			_srcFolder = Path.Combine(Path.GetTempPath(), "~!source");
> ```

You could add an additional random part so that multiple builds can run in parallel on a build agent, e.g.

    var tempDir = Path.Combine(Path.GetTempPath(), Path.GetRandomFileName());
    _srcFolder = Path.Combine(tempDir, "~!source");

---

*[SIL.Core.Desktop.Tests/IO/DirectoryUtilitiesTests.cs, line 166 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/622#-L6LoRky9OsG8lWjCAdP:-L6LoRky9OsG8lWjCAdQ:b-by2v8o) ([raw file](https://github.com/sillsdev/libpalaso/blob/9eac80d73b6633abd88543988053b9e72221ad93/SIL.Core.Desktop.Tests/IO/DirectoryUtilitiesTests.cs#L166)):*
> ```Smalltalk
> 		/// ------------------------------------------------------------------------------------
> 		[Test]
> 		public void CopyFolder_DestinationFolderDoesNotExist_ReturnsFalse()
> ```

It might be better to start the test name with the name of the method we're testing, e.g. `CopyDirectory_DestinationFolderDoesNotExist_ReturnsFalse()`. There is a similar test towards the top of the file that tests `CopyDirectoryContents`. If both tests start with `CopyFolder_` it's harder to see why there are two and why they are different.

---

*[SIL.Core.Desktop.Tests/IO/FileLocatorTests.cs, line 6 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/622#-L6Lp4jIB21LbNbrCmU7:-L6Lp4jJ_x2Yf7eC-Nzb:bbvb305) ([raw file](https://github.com/sillsdev/libpalaso/blob/9eac80d73b6633abd88543988053b9e72221ad93/SIL.Core.Desktop.Tests/IO/FileLocatorTests.cs#L6)):*
> ```Smalltalk
> using SIL.Reporting;
> using System;
> using System.IO;
> ```

Usually we have the system assemblies at the beginning of the list - that's also the order in which R# and similar tools sort the usings.

---

*[SIL.Core.Desktop.Tests/IO/PathUtilitiesTests.cs, line 126 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/622#-L6LwpOunl5eGNQARqaw:-L6LwpOunl5eGNQARqax:bav5pp5) ([raw file](https://github.com/sillsdev/libpalaso/blob/9eac80d73b6633abd88543988053b9e72221ad93/SIL.Core.Desktop.Tests/IO/PathUtilitiesTests.cs#L126)):*
> ```Smalltalk
> 		public void SelectFileInExplorer_PathHasCombiningCharacters_StillOpensAndSelects()
> 		{
> 			var path = Path.Combine(Path.GetTempPath(), "ปู  should select this file");
> ```

IMO it would be better to add a random component to the filename and then always create it. Also, this test doesn't clean up afterwards.

---

*[SIL.Core.Desktop.Tests/IO/PathUtilitiesTests.cs, line 137 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/622#-L6LxFYgKtZViFTA-2pQ:-L6LxFYhpurZm3WZEpMh:bbw4h1b) ([raw file](https://github.com/sillsdev/libpalaso/blob/9eac80d73b6633abd88543988053b9e72221ad93/SIL.Core.Desktop.Tests/IO/PathUtilitiesTests.cs#L137)):*
> ```Smalltalk
> 		public void SelectFileInExplorer_PathIsADirectoryHasCombiningCharacters_StillOpensAndSelects()
> 		{
> 			var path = Path.Combine(Path.GetTempPath(), "ปู  should select this directory");
> ```

see above

---

*[SIL.Core.Desktop.Tests/IO/PathUtilitiesTests.cs, line 151 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/622#-L6LxKggMYU1rPriJY7f:-L6LxKggMYU1rPriJY7g:byr4ffv) ([raw file](https://github.com/sillsdev/libpalaso/blob/9eac80d73b6633abd88543988053b9e72221ad93/SIL.Core.Desktop.Tests/IO/PathUtilitiesTests.cs#L151)):*
> ```Smalltalk
> 			//as of May 27 2015, this is expected to fail on Windows. See enhancment note 
> 			//in the OpenDirectoryInExplorer() code.
> 			var path = Path.Combine(Path.GetTempPath(), "ปู should select this directory");
> ```

see above

---

*[SIL.Core.Desktop/IO/DirectoryUtilities.cs, line 25 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/622#-L6MQvmAQPR0keuVfUk0:-L6MQvmAQPR0keuVfUk1:b99yt4n) ([raw file](https://github.com/sillsdev/libpalaso/blob/9eac80d73b6633abd88543988053b9e72221ad93/SIL.Core.Desktop/IO/DirectoryUtilities.cs#L25)):*
> ```Smalltalk
> 		/// <param name="srcDirectory">Directory to copy</param>
> 		/// <returns>Null if the copy was unsuccessful, otherwise the path to the copied directory</returns>
> ```

nit-pick: extra empty line

---

*[SIL.Core.Desktop/IO/DirectoryUtilities.cs, line 31 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/622#-L6MQxwzXs1-83UL3qhN:-L6MQxwzXs1-83UL3qhO:b99yt4n) ([raw file](https://github.com/sillsdev/libpalaso/blob/9eac80d73b6633abd88543988053b9e72221ad93/SIL.Core.Desktop/IO/DirectoryUtilities.cs#L31)):*
> ```Smalltalk
> 			return (CopyDirectory(srcDirectory, Path.GetTempPath(), out dstDirectory) ? dstDirectory : null);
> 		}
> ```

nit-pick: extra empty line

---

*[SIL.Core.Desktop/IO/DirectoryUtilities.cs, line 44 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/622#-L6MQyiAxNHQFEQ1retR:-L6MQyiAxNHQFEQ1retS:b99yt4n) ([raw file](https://github.com/sillsdev/libpalaso/blob/9eac80d73b6633abd88543988053b9e72221ad93/SIL.Core.Desktop/IO/DirectoryUtilities.cs#L44)):*
> ```Smalltalk
> 		/// <param name="dstDirectoryParent">Destination directory where source directory and its contents are copied</param>
> 		/// <returns>true if successful, otherwise, false.</returns>
> ```

nit-pick: extra empty line

---

*[SIL.Core.Desktop/IO/DirectoryUtilities.cs, line 51 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/622#-L6MQzPDi5NzvNrE53kl:-L6MQzPDi5NzvNrE53km:b99yt4n) ([raw file](https://github.com/sillsdev/libpalaso/blob/9eac80d73b6633abd88543988053b9e72221ad93/SIL.Core.Desktop/IO/DirectoryUtilities.cs#L51)):*
> ```Smalltalk
> 		}
> ```

nit-pick: extra empty line

---

*[SIL.Core.Desktop/IO/DirectoryUtilities.cs, line 72 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/622#-L6MR-5aisd4nDqxiSLJ:-L6MR-5aisd4nDqxiSLK:b99yt4n) ([raw file](https://github.com/sillsdev/libpalaso/blob/9eac80d73b6633abd88543988053b9e72221ad93/SIL.Core.Desktop/IO/DirectoryUtilities.cs#L72)):*
> ```Smalltalk
> 		}
> ```

nit-pick: extra empty line

---

*[SIL.Core.Desktop/IO/DirectoryUtilities.cs, line 76 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/622#-L6MPz6oyTLhsfKAzVSg:-L6MPz6oyTLhsfKAzVSh:b-ah91a4) ([raw file](https://github.com/sillsdev/libpalaso/blob/9eac80d73b6633abd88543988053b9e72221ad93/SIL.Core.Desktop/IO/DirectoryUtilities.cs#L76)):*
> directorys

s/directorys/directories/

---

*[SIL.Core.Desktop/IO/DirectoryUtilities.cs, line 124 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/622#-L6MQU11EpjhKpi6CAAm:-L6MQU11EpjhKpi6CAAn:b-kwrk8z) ([raw file](https://github.com/sillsdev/libpalaso/blob/9eac80d73b6633abd88543988053b9e72221ad93/SIL.Core.Desktop/IO/DirectoryUtilities.cs#L124)):*
> ```Smalltalk
> 		/// <param name="showErrorMessage"></param>
> 		/// <returns>True if able to set access, False otherwise</returns>
> 		public static bool SetFullControl(string fullDirectoryPath, bool showErrorMessage = true)
> ```

Does this work on Linux?

---

*[SIL.Core.Tests/IO/DirectoryHelperTests.cs, line 177 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/622#-L6MT1_uyEjHpwVSilHq:-L6MT1_uyEjHpwVSilHr:b-cc6ybm) ([raw file](https://github.com/sillsdev/libpalaso/blob/9eac80d73b6633abd88543988053b9e72221ad93/SIL.Core.Tests/IO/DirectoryHelperTests.cs#L177)):*
> ```Smalltalk
> 			catch
> 			{
> 				// Ignore -- can't test this on this system
> ```

Assert.Ignore("Can't test this on this system");

---

*[SIL.Core.Tests/IO/DirectoryHelperTests.cs, line 196 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/622#-L6MTJ33YRPnZMT3Rg0d:-L6MTJ33YRPnZMT3Rg0e:b-umkuw7) ([raw file](https://github.com/sillsdev/libpalaso/blob/9eac80d73b6633abd88543988053b9e72221ad93/SIL.Core.Tests/IO/DirectoryHelperTests.cs#L196)):*
> ```Smalltalk
> 					try
> 					{
> 						Directory.SetCurrentDirectory(Path.Combine(logicalDrive, folder));
> ```

Looks like we should break out of the loop if we successfully set the current dir

---

*[SIL.Core.Tests/IO/DirectoryHelperTests.cs, line 204 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/622#-L6MUeX4DmFe1zQHHui9:-L6MUeX4DmFe1zQHHuiA:b-n3tw6t) ([raw file](https://github.com/sillsdev/libpalaso/blob/9eac80d73b6633abd88543988053b9e72221ad93/SIL.Core.Tests/IO/DirectoryHelperTests.cs#L204)):*
> ```Smalltalk
> 				}
> 				// 02 SEP 2013, Phil Hopper: set correct directory separator for OS
> 				Assert.IsTrue(DirectoryHelper.AreEquivalent(_directorySeparator, logicalDrive));
> ```

This test method doesn't make much sense on Linux - _directorySeparator will always be the root directory. Maybe we should exclude it on Linux.

---

*[SIL.Core.Tests/IO/FileHelperTests.cs, line 55 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/622#-L6MWY_YnXKCYmaqCmfC:-L6MWY_YnXKCYmaqCmfD:b-kwrk8z) ([raw file](https://github.com/sillsdev/libpalaso/blob/9eac80d73b6633abd88543988053b9e72221ad93/SIL.Core.Tests/IO/FileHelperTests.cs#L55)):*
> ```Smalltalk
> 				try
> 				{
> 					Assert.IsTrue(FileHelper.IsLocked(file.Path));
> ```

Does this work on Linux?

---

*[SIL.Windows.Forms.Tests/SettingsProviderTests.cs, line 154 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/622#-L6MbEIoikMUYuPTZueZ:-L6MbEIpTmaPwOvl-z2s:be7q020) ([raw file](https://github.com/sillsdev/libpalaso/blob/9eac80d73b6633abd88543988053b9e72221ad93/SIL.Windows.Forms.Tests/SettingsProviderTests.cs#L154)):*
> ```Smalltalk
> 
> 		[Test]
> 		[Category("KnownMonoIssue")]
> ```

Can you add a comment what the problem is? IRC it's the buggy settings implementation that we fixed in our mono*-sil, but that hasn't made it in an upstream release yet.

---

*[SIL.WritingSystems/IcuRulesCollator.cs, line 59 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/622#-L6MaCDffX8LkOoGewsO:-L6MaCDgV-tX2SGZTSkE:b-7h33z5) ([raw file](https://github.com/sillsdev/libpalaso/blob/9eac80d73b6633abd88543988053b9e72221ad93/SIL.WritingSystems/IcuRulesCollator.cs#L59)):*
> ```Smalltalk
> 			{
> 				keyDataFieldName = "key";
> 				origStringFieldName = "source";
> ```

I'm confused. Why do we need this distinction? Where is the implementation? If it's in icu.net we should fix it there instead of complicating the code here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/sillsdev/libpalaso/622)*
<!-- Sent from Reviewable.io -->

ermshiperete commented:
Some remarks, questions and suggestions, but overall <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 36 of 188 files at r1.
Review status: all files reviewed at latest revision, 27 unresolved discussions, some commit checks failed.

---

*[SIL.Core.Desktop.Tests/IO/DirectoryUtilitiesTests.cs, line 1 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/622#-L6N7-9_wXv4Hbb6bcRF:-L6N7-9_wXv4Hbb6bcRG:b-nytc1s) ([raw file](https://github.com/sillsdev/libpalaso/blob/9eac80d73b6633abd88543988053b9e72221ad93/SIL.Core.Desktop.Tests/IO/DirectoryUtilitiesTests.cs#L1)):*
> ```Smalltalk
> using NUnit.Framework;
> ```

nit-pick: Missing file header - same for several other files.

---

*[SIL.Core/IO/DirectoryHelper.cs, line 76 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/622#-L6N2MMxd9VvGFZoRw-V:-L6N2MMyIccNbGhK5wrJ:bq5mj2t) ([raw file](https://github.com/sillsdev/libpalaso/blob/9eac80d73b6633abd88543988053b9e72221ad93/SIL.Core/IO/DirectoryHelper.cs#L76)):*
> ```Smalltalk
> 			var comparison = Platform.IsWindows ? StringComparison.InvariantCultureIgnoreCase : StringComparison.InvariantCulture;
> 			var backslash = new char[] { '\\' }; // added this step because mono does not implicitly convert from char to char[]
> 			return string.Compare(dirInfo1.FullName.TrimEnd(backslash), dirInfo2.FullName.TrimEnd(backslash), comparison) == 0;
> ```

I wonder if this works on  Linux if one of the two arguments ends in a forward slash. Do we have a test for that?

---

*[SIL.Core/IO/FileHelper.cs, line 37 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/622#-L6N31_L5M8qwkIBJEtx:-L6N31_M8e3TToEHlU-X:bgozxgf) ([raw file](https://github.com/sillsdev/libpalaso/blob/9eac80d73b6633abd88543988053b9e72221ad93/SIL.Core/IO/FileHelper.cs#L37)):*
> ```Smalltalk
> 					}
> 				}
> 				reader.Close();
> ```

If we find a match we don't call Close(), but I guess it's implicit because of the surrounding `using`. I suggest removing this line.

---

*[SIL.Core/PlatformUtilities/Platform.cs, line 14 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/622#-L6N5s8pNu8jGsSSXv4J:-L6N5s8pNu8jGsSSXv4K:buh9vt4) ([raw file](https://github.com/sillsdev/libpalaso/blob/9eac80d73b6633abd88543988053b9e72221ad93/SIL.Core/PlatformUtilities/Platform.cs#L14)):*
> ```Smalltalk
> 		private static string _sessionManager;
> 
> 		public static bool IsUnix => Environment.OSVersion.Platform == PlatformID.Unix;
> ```

@chrisvire Since I think you wrote the original code that made the distinction  between Linux and Mac, maybe you want to take a look at this change if this works as expected on Mac.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/sillsdev/libpalaso/622#-:-L6N7A23wmOycxtMPgJl:b-5jg76j)*
<!-- Sent from Reviewable.io -->

jasonleenaylor commented on ExtractCopyright.Tests/ExtractCopyrightTests.cs:
```diff
@@ -1,4 +1,4 @@
-﻿// Copyright (c) 2017 SIL International
+// Copyright (c) 2017 SIL International
```
2018

jasonleenaylor commented on ExtractCopyright.Tests/ExtractCopyrightTests.cs:
```diff
@@ -178,7 +178,7 @@ public void CreateCopyrightFileWorks()
 			}
 
-			Assert.LessOrEqual(ackDict.Keys.Count + 3, copyrights.Paragraphs.Count, "We should have the two standard paragraphs, one paragraph per acknowledgement, and at least one license paragraph");
+			Assert.LessOrEqual(ackDict.Keys.Count - 1 + 3, copyrights.Paragraphs.Count, "We should have the two standard paragraphs, one paragraph per acknowledgement, and at least one license paragraph");
```
This extra magic number needs some explanation. The reason isn't clear. Perhaps it shouldn't have used any magic numbers to begin with, but this isn't helping.

jasonleenaylor commented on SIL.BuildTasks/StampAssemblies/StampAssemblies.cs:
```diff
@@ -152,38 +190,64 @@ public VersionParts GetExistingAssemblyFileVersion(string contents)
 		}
 
-		public VersionParts ParseVersionString(string contents, bool allowHashAsRevision = false)
+		public static VersionParts ParseVersionString(string contents, VersionFormat format = VersionFormat.File)
```
API change here, Are we going to 5.0 for the library now?

ddaspit replied:
I think this API was only public for unit test purposes. If projects are using it, I can keep the old API as well.

jasonleenaylor commented on SIL.Core.Desktop.Tests/IO/FileLocatorTests.cs:
```diff
@@ -1,9 +1,9 @@
 using SIL.PlatformUtilities;
 using SIL.Reporting;
+using System;
+using System.IO;
```
These re-orderings are counter to our current best practices (R# encouraged)

jasonleenaylor commented on SIL.Core.Desktop.Tests/IO/PathUtilitiesTests.cs:
```diff
@@ -0,0 +1,160 @@
+// Copyright (c) 2014 SIL International
```
2014-2018

jasonleenaylor commented on SIL.Core.Tests/Email/MAPIHelperTests.cs:
```diff
@@ -0,0 +1,18 @@
+using NUnit.Framework;
```
license header

jasonleenaylor commented on SIL.Core.Tests/IO/FileHelperTests.cs:
```diff
@@ -0,0 +1,83 @@
+using NUnit.Framework;
```
header

jasonleenaylor commented on SIL.Core.Tests/IO/FileTestEnvironment.cs:
```diff
@@ -0,0 +1,21 @@
+using SIL.IO;
```
header

jasonleenaylor commented on SIL.Core.Tests/IO/RobustFileTests.cs:
```diff
@@ -1,11 +1,8 @@
-using System.Xml;
 using NUnit.Framework;
 using SIL.IO;
+using System;
```
more re-ordering

jasonleenaylor commented on SIL.Core/CommandLineProcessing/CommandLineRunner.cs:
```diff
@@ -1,4 +1,4 @@
-﻿using System;
+using System;
```
header

jasonleenaylor commented on SIL.Core/IO/DirectoryHelper.cs:
```diff
@@ -0,0 +1,114 @@
+using SIL.PlatformUtilities;
```
header

jasonleenaylor commented on SIL.Core/IO/FileHelper.cs:
```diff
@@ -0,0 +1,42 @@
+using System.IO;
```
header

jasonleenaylor commented on SIL.Core/IO/PathHelper.cs:
```diff
@@ -0,0 +1,219 @@
+using SIL.PlatformUtilities;
```
header

jasonleenaylor commented on SIL.Core/IO/RobustIO.cs:
```diff
@@ -1,8 +1,11 @@
-﻿using System.IO;
+using System;
```
header

jasonleenaylor commented on SIL.Core/IO/TempFile.cs:
```diff
@@ -1,8 +1,8 @@
 using System.IO;
 using System.Text;
-using SIL.Reporting;
+//using SIL.Reporting;
```
remove

jasonleenaylor commented on SIL.Core/IO/TempFile.cs:
```diff
@@ -1,8 +1,8 @@
-﻿using System;
+using System;
```
header

jasonleenaylor commented on SIL.Core/Progress/XslTransformWithProgress.cs:
```diff
@@ -109,18 +109,6 @@ public void Transform(ProgressState progressState)
 					_progressState.WriteToLog(err.Message);
 					_progressState.State = ProgressState.StateValue.StoppedWithError;
 				}
-				finally
```
So transforms clean up their own temp files now?

ddaspit replied:
I'm not sure. This TemporaryFiles property was removed in .NET Core and .NET Standard, so there is no way to cleanup the files.

ddaspit replied:
I added the code back with conditional compilation.

jasonleenaylor commented on SIL.Core/Properties/AssemblyInfo.cs:
```diff
@@ -1,45 +1,20 @@
-	Url = "http://www.codeplex.com/DotNetZip",
-	LicenseUrl = "https://opensource.org/licenses/MS-PL", Location = "./Ionic.Zip.dll",
-	Html = "<li><a href='http://www.codeplex.com/DotNetZip'>Ionic.Zip</a> © Dino Chiesa 2006-2011 (<a href='https://opensource.org/licenses/MS-PL'>MS-PL</a>) - a library for handling zip archives (Flavor=Retail)</li>")]
-[assembly: Acknowledgement("NDesk.DBus", Name = "NDesk.DBus", Url = "https://www.nuget.org/packages/NDesk.DBus/",
```
Should be replaced by one for Posix i'm guessing?

ddaspit replied:
There is already an acknowledgement for Mono.Posix.

jasonleenaylor commented on SIL.Core/SIL.Core.csproj:
```diff
@@ -1,585 +1,29 @@
-      <DesignTime>True</DesignTime>
-      <AutoGen>True</AutoGen>
-    </Compile>
-    <Compile Include="Properties\AssemblyInfo.cs" />
```
ok, so I asked @andrew-polk about this on his PR. How does this work now and why won't I run into problems if I'm changing branches which have different sets of files?

jasonleenaylor replied:
So given the discussion on the other PR I'm happy about this change now.

ddaspit commented:
Review status: all files reviewed at latest revision, 43 unresolved discussions, some commit checks failed.

---

*[SIL.BuildTasks/StampAssemblies/StampAssemblies.cs, line 193 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/622#-L6KXUrPmNz3BcxL53Id-r1-193:-L6o6J5MU1zdVamwvdbZ:b-7krai2) ([raw file](https://github.com/sillsdev/libpalaso/blob/9eac80d73b6633abd88543988053b9e72221ad93/SIL.BuildTasks/StampAssemblies/StampAssemblies.cs#L193)):*
<details><summary><i>Previously, jasonleenaylor (Jason Naylor) wrote…</i></summary><blockquote>

API change here, Are we going to 5.0 for the library now?
</blockquote></details>

This is a static method on a build task class, so I doubt it is being called by anyone directly. I think it is public to facility unit testing.

---

*[SIL.Core.Desktop.Tests/IO/FileLocatorTests.cs, line 6 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/622#-L6Lp4jIB21LbNbrCmU7:-L6o8xuDRP1axxLIsSy0:b-fnnf9k) ([raw file](https://github.com/sillsdev/libpalaso/blob/9eac80d73b6633abd88543988053b9e72221ad93/SIL.Core.Desktop.Tests/IO/FileLocatorTests.cs#L6)):*
<details><summary><i>Previously, ermshiperete (Eberhard Beilharz) wrote…</i></summary><blockquote>

Usually we have the system assemblies at the beginning of the list - that's also the order in which R# and similar tools sort the usings.
</blockquote></details>

I haven't been using R# lately (slows down VS too much, and VS 2017 seems good enough), so I was using the "Remove and Sort Usings" command in VS 2017. By default, it sorts usings without putting system assemblies first. I found a setting in VS to change this behavior and sort system assemblies first.

---

*[SIL.Core.Desktop.Tests/IO/PathUtilitiesTests.cs, line 126 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/622#-L6LwpOunl5eGNQARqaw:-L6oEO01EeoLVLdDXvIt:b-wiksp1) ([raw file](https://github.com/sillsdev/libpalaso/blob/9eac80d73b6633abd88543988053b9e72221ad93/SIL.Core.Desktop.Tests/IO/PathUtilitiesTests.cs#L126)):*
<details><summary><i>Previously, ermshiperete (Eberhard Beilharz) wrote…</i></summary><blockquote>

IMO it would be better to add a random component to the filename and then always create it. Also, this test doesn't clean up afterwards.
</blockquote></details>

I think this is done on purpose. This test requires that a developer visually inspect the results to see if it worked.

---

*[SIL.Core.Desktop.Tests/IO/PathUtilitiesTests.cs, line 137 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/622#-L6LxFYgKtZViFTA-2pQ:-L6oEdcOLKpf79JAR4aG:byr4ffv) ([raw file](https://github.com/sillsdev/libpalaso/blob/9eac80d73b6633abd88543988053b9e72221ad93/SIL.Core.Desktop.Tests/IO/PathUtilitiesTests.cs#L137)):*
<details><summary><i>Previously, ermshiperete (Eberhard Beilharz) wrote…</i></summary><blockquote>

see above
</blockquote></details>

see above

---

*[SIL.Core.Desktop.Tests/IO/PathUtilitiesTests.cs, line 151 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/622#-L6LxKggMYU1rPriJY7f:-L6oEfigl1h2XPF5ercp:byr4ffv) ([raw file](https://github.com/sillsdev/libpalaso/blob/9eac80d73b6633abd88543988053b9e72221ad93/SIL.Core.Desktop.Tests/IO/PathUtilitiesTests.cs#L151)):*
<details><summary><i>Previously, ermshiperete (Eberhard Beilharz) wrote…</i></summary><blockquote>

see above
</blockquote></details>

see above

---

*[SIL.Core.Desktop/IO/DirectoryUtilities.cs, line 124 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/622#-L6MQU11EpjhKpi6CAAm:-L6oG03xbN4DXTT46CZR:brfk71u) ([raw file](https://github.com/sillsdev/libpalaso/blob/9eac80d73b6633abd88543988053b9e72221ad93/SIL.Core.Desktop/IO/DirectoryUtilities.cs#L124)):*
<details><summary><i>Previously, ermshiperete (Eberhard Beilharz) wrote…</i></summary><blockquote>

Does this work on Linux?
</blockquote></details>

Probably not. I added a platform check at the beginning of the method.

---

*[SIL.Core.Tests/IO/DirectoryHelperTests.cs, line 204 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/622#-L6MUeX4DmFe1zQHHui9:-L6oHTm7AxGK5AX_QyRi:b-t1smuv) ([raw file](https://github.com/sillsdev/libpalaso/blob/9eac80d73b6633abd88543988053b9e72221ad93/SIL.Core.Tests/IO/DirectoryHelperTests.cs#L204)):*
<details><summary><i>Previously, ermshiperete (Eberhard Beilharz) wrote…</i></summary><blockquote>

This test method doesn't make much sense on Linux - _directorySeparator will always be the root directory. Maybe we should exclude it on Linux.
</blockquote></details>

I'm not sure. This was an existing test that I moved here. I think I'll just leave it as is.

---

*[SIL.Core.Tests/IO/FileHelperTests.cs, line 55 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/622#-L6MWY_YnXKCYmaqCmfC:-L6oHmHhVw9KG454uosc:bp0t96x) ([raw file](https://github.com/sillsdev/libpalaso/blob/9eac80d73b6633abd88543988053b9e72221ad93/SIL.Core.Tests/IO/FileHelperTests.cs#L55)):*
<details><summary><i>Previously, ermshiperete (Eberhard Beilharz) wrote…</i></summary><blockquote>

Does this work on Linux?
</blockquote></details>

I'm not sure. This is an existing test that I moved from here. 

---

*[SIL.WritingSystems/IcuRulesCollator.cs, line 59 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/622#-L6MaCDffX8LkOoGewsO:-L6oJPxzx4cc72UZJccG:b-e38570) ([raw file](https://github.com/sillsdev/libpalaso/blob/9eac80d73b6633abd88543988053b9e72221ad93/SIL.WritingSystems/IcuRulesCollator.cs#L59)):*
<details><summary><i>Previously, ermshiperete (Eberhard Beilharz) wrote…</i></summary><blockquote>

I'm confused. Why do we need this distinction? Where is the implementation? If it's in icu.net we should fix it there instead of complicating the code here.
</blockquote></details>

This is a horrible hack that is required, because we cannot initialize SortKey objects directly. We have to set the internal state through reflection. This used to be handled by ICU.NET, but in 2.3 it was removed for the .NET Standard build, because it is targeting .NET Standard 1.6, which does not have the SortKey class. The .NET Standard build of ICU.NET had a different API than the .NET Framework build. It returns its own SortKey class. SortKey is now in .NET Standard 2.0, so we could update ICU.NET to target .NET Standard 2.0 and put the code back. At that point, we could remove the code here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/sillsdev/libpalaso/622)*
<!-- Sent from Reviewable.io -->

ddaspit commented:
Review status: all files reviewed at latest revision, 43 unresolved discussions, some commit checks failed.

---

*[NuGet.config, line 4 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/622#-L6LigHm4eQ183VZb1qT:-L6oODFtjx9vXrpVdZ_5:b-896fix) ([raw file](https://github.com/sillsdev/libpalaso/blob/9eac80d73b6633abd88543988053b9e72221ad93/NuGet.config#L4)):*
<details><summary><i>Previously, ermshiperete (Eberhard Beilharz) wrote…</i></summary><blockquote>

It would be better to use https
</blockquote></details>

Done.

---

*[build/NuGet.targets, line 14 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/622#-L6Lj8ij6FPezdFsequG:-L6oOEZpAor8wbHW4QXw:b-896fix) ([raw file](https://github.com/sillsdev/libpalaso/blob/9eac80d73b6633abd88543988053b9e72221ad93/build/NuGet.targets#L14)):*
<details><summary><i>Previously, ermshiperete (Eberhard Beilharz) wrote…</i></summary><blockquote>

nit-pick: indent with tabs instead of spaces
</blockquote></details>

Done.

---

*[ExtractCopyright.Tests/ExtractCopyrightTests.cs, line 1 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/622#-L6KXUrPmNz3BcxL53IS-r1-1:-L6oOFYRjFTBBG0yJX2q:b-896fix) ([raw file](https://github.com/sillsdev/libpalaso/blob/9eac80d73b6633abd88543988053b9e72221ad93/ExtractCopyright.Tests/ExtractCopyrightTests.cs#L1)):*
<details><summary><i>Previously, jasonleenaylor (Jason Naylor) wrote…</i></summary><blockquote>

2018
</blockquote></details>

Done.

---

*[ExtractCopyright.Tests/ExtractCopyrightTests.cs, line 181 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/622#-L6KXUrPmNz3BcxL53IS-r1-181:-L6oOGYWdVqId9EZgUP2:b-896fix) ([raw file](https://github.com/sillsdev/libpalaso/blob/9eac80d73b6633abd88543988053b9e72221ad93/ExtractCopyright.Tests/ExtractCopyrightTests.cs#L181)):*
<details><summary><i>Previously, jasonleenaylor (Jason Naylor) wrote…</i></summary><blockquote>

This extra magic number needs some explanation. The reason isn't clear. Perhaps it shouldn't have used any magic numbers to begin with, but this isn't helping.
</blockquote></details>

Done.

---

*[SIL.BuildTasks/StampAssemblies/StampAssemblies.cs, line 151 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/622#-L6LlWk3VMr9aa0lgZcj:-L6oOHYS_frr6RUzIrI4:b-896fix) ([raw file](https://github.com/sillsdev/libpalaso/blob/9eac80d73b6633abd88543988053b9e72221ad93/SIL.BuildTasks/StampAssemblies/StampAssemblies.cs#L151)):*
<details><summary><i>Previously, ermshiperete (Eberhard Beilharz) wrote…</i></summary><blockquote>

Could write this as

    Prerelease = incoming.Prerelease ?? existing.Prerelease

if we don't want the `if` below.
</blockquote></details>

Done.

---

*[SIL.Core/SIL.Core.csproj, line 1 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/622#-L6LmMVSozFF8K1wju2w:-L6oOMWGUB5WQtOZ65gA:bp8m3ey) ([raw file](https://github.com/sillsdev/libpalaso/blob/9eac80d73b6633abd88543988053b9e72221ad93/SIL.Core/SIL.Core.csproj#L1)):*
<details><summary><i>Previously, ermshiperete (Eberhard Beilharz) wrote…</i></summary><blockquote>

So much shorter. Nice! :smile: 
</blockquote></details>

Agreed. The new csproj format is much nicer.

---

*[SIL.Core.Desktop.Tests/IO/DirectoryUtilitiesTests.cs, line 1 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/622#-L6N7-9_wXv4Hbb6bcRF:-L6oOS39ozUqLvtBvN1g:b-896fix) ([raw file](https://github.com/sillsdev/libpalaso/blob/9eac80d73b6633abd88543988053b9e72221ad93/SIL.Core.Desktop.Tests/IO/DirectoryUtilitiesTests.cs#L1)):*
<details><summary><i>Previously, ermshiperete (Eberhard Beilharz) wrote…</i></summary><blockquote>

nit-pick: Missing file header - same for several other files.
</blockquote></details>

Done.

---

*[SIL.Core.Desktop.Tests/IO/DirectoryUtilitiesTests.cs, line 19 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/622#-L6LmlfVqKdzCOpMkplg:-L6oOT8m7735uGy3twue:b-896fix) ([raw file](https://github.com/sillsdev/libpalaso/blob/9eac80d73b6633abd88543988053b9e72221ad93/SIL.Core.Desktop.Tests/IO/DirectoryUtilitiesTests.cs#L19)):*
<details><summary><i>Previously, ermshiperete (Eberhard Beilharz) wrote…</i></summary><blockquote>

You could add an additional random part so that multiple builds can run in parallel on a build agent, e.g.

    var tempDir = Path.Combine(Path.GetTempPath(), Path.GetRandomFileName());
    _srcFolder = Path.Combine(tempDir, "~!source");
</blockquote></details>

Done.

---

*[SIL.Core.Desktop.Tests/IO/DirectoryUtilitiesTests.cs, line 166 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/622#-L6LoRky9OsG8lWjCAdP:-L6oOTplHS28Zb3kVdj0:b-896fix) ([raw file](https://github.com/sillsdev/libpalaso/blob/9eac80d73b6633abd88543988053b9e72221ad93/SIL.Core.Desktop.Tests/IO/DirectoryUtilitiesTests.cs#L166)):*
<details><summary><i>Previously, ermshiperete (Eberhard Beilharz) wrote…</i></summary><blockquote>

It might be better to start the test name with the name of the method we're testing, e.g. `CopyDirectory_DestinationFolderDoesNotExist_ReturnsFalse()`. There is a similar test towards the top of the file that tests `CopyDirectoryContents`. If both tests start with `CopyFolder_` it's harder to see why there are two and why they are different.
</blockquote></details>

Done.

---

*[SIL.Core.Desktop.Tests/IO/FileLocatorTests.cs, line 6 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/622#-L6KXUrQMvHvvkTWD8mf-r1-6:-L6oOUUyorMVyvjHy2yH:b-896fix) ([raw file](https://github.com/sillsdev/libpalaso/blob/9eac80d73b6633abd88543988053b9e72221ad93/SIL.Core.Desktop.Tests/IO/FileLocatorTests.cs#L6)):*
<details><summary><i>Previously, jasonleenaylor (Jason Naylor) wrote…</i></summary><blockquote>

These re-orderings are counter to our current best practices (R# encouraged)
</blockquote></details>

Done.

---

*[SIL.Core.Desktop.Tests/IO/PathUtilitiesTests.cs, line 1 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/622#-L6KXUrQMvHvvkTWD8mh-r1-1:-L6oOX0n6O8rhSXJEGch:b-896fix) ([raw file](https://github.com/sillsdev/libpalaso/blob/9eac80d73b6633abd88543988053b9e72221ad93/SIL.Core.Desktop.Tests/IO/PathUtilitiesTests.cs#L1)):*
<details><summary><i>Previously, jasonleenaylor (Jason Naylor) wrote…</i></summary><blockquote>

2014-2018
</blockquote></details>

Done.

---

*[SIL.Core.Desktop/IO/DirectoryUtilities.cs, line 25 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/622#-L6MQvmAQPR0keuVfUk0:-L6oOZCLbNcvmSUizTmM:b-896fix) ([raw file](https://github.com/sillsdev/libpalaso/blob/9eac80d73b6633abd88543988053b9e72221ad93/SIL.Core.Desktop/IO/DirectoryUtilities.cs#L25)):*
<details><summary><i>Previously, ermshiperete (Eberhard Beilharz) wrote…</i></summary><blockquote>

nit-pick: extra empty line
</blockquote></details>

Done.

---

*[SIL.Core.Desktop/IO/DirectoryUtilities.cs, line 31 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/622#-L6MQxwzXs1-83UL3qhN:-L6oOZnTSEvOcUknnzxv:b-896fix) ([raw file](https://github.com/sillsdev/libpalaso/blob/9eac80d73b6633abd88543988053b9e72221ad93/SIL.Core.Desktop/IO/DirectoryUtilities.cs#L31)):*
<details><summary><i>Previously, ermshiperete (Eberhard Beilharz) wrote…</i></summary><blockquote>

nit-pick: extra empty line
</blockquote></details>

Done.

---

*[SIL.Core.Desktop/IO/DirectoryUtilities.cs, line 44 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/622#-L6MQyiAxNHQFEQ1retR:-L6oP1_Xu-q9-O3N3wwO:b-896fix) ([raw file](https://github.com/sillsdev/libpalaso/blob/9eac80d73b6633abd88543988053b9e72221ad93/SIL.Core.Desktop/IO/DirectoryUtilities.cs#L44)):*
<details><summary><i>Previously, ermshiperete (Eberhard Beilharz) wrote…</i></summary><blockquote>

nit-pick: extra empty line
</blockquote></details>

Done.

---

*[SIL.Core.Desktop/IO/DirectoryUtilities.cs, line 51 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/622#-L6MQzPDi5NzvNrE53kl:-L6oO_dQg-j-oEkiF4j3:b-896fix) ([raw file](https://github.com/sillsdev/libpalaso/blob/9eac80d73b6633abd88543988053b9e72221ad93/SIL.Core.Desktop/IO/DirectoryUtilities.cs#L51)):*
<details><summary><i>Previously, ermshiperete (Eberhard Beilharz) wrote…</i></summary><blockquote>

nit-pick: extra empty line
</blockquote></details>

Done.

---

*[SIL.Core.Desktop/IO/DirectoryUtilities.cs, line 72 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/622#-L6MR-5aisd4nDqxiSLJ:-L6oOaG7qE0FjWpXOuNB:b-896fix) ([raw file](https://github.com/sillsdev/libpalaso/blob/9eac80d73b6633abd88543988053b9e72221ad93/SIL.Core.Desktop/IO/DirectoryUtilities.cs#L72)):*
<details><summary><i>Previously, ermshiperete (Eberhard Beilharz) wrote…</i></summary><blockquote>

nit-pick: extra empty line
</blockquote></details>

Done.

---

*[SIL.Core.Desktop/IO/DirectoryUtilities.cs, line 76 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/622#-L6MPz6oyTLhsfKAzVSg:-L6oOagHW1u5I8L62lR2:b-896fix) ([raw file](https://github.com/sillsdev/libpalaso/blob/9eac80d73b6633abd88543988053b9e72221ad93/SIL.Core.Desktop/IO/DirectoryUtilities.cs#L76)):*
<details><summary><i>Previously, ermshiperete (Eberhard Beilharz) wrote…</i></summary><blockquote>

> directorys

s/directorys/directories/
</blockquote></details>

Done.

---

*[SIL.Core.Tests/Email/MAPIHelperTests.cs, line 1 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/622#-L6KXUrQMvHvvkTWD8nH-r1-1:-L6oOc6Djf8ep02iACxd:b-896fix) ([raw file](https://github.com/sillsdev/libpalaso/blob/9eac80d73b6633abd88543988053b9e72221ad93/SIL.Core.Tests/Email/MAPIHelperTests.cs#L1)):*
<details><summary><i>Previously, jasonleenaylor (Jason Naylor) wrote…</i></summary><blockquote>

license header
</blockquote></details>

Done.

---

*[SIL.Core.Tests/IO/DirectoryHelperTests.cs, line 177 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/622#-L6MT1_uyEjHpwVSilHq:-L6oOd2FBLAPgZu53kVL:b-896fix) ([raw file](https://github.com/sillsdev/libpalaso/blob/9eac80d73b6633abd88543988053b9e72221ad93/SIL.Core.Tests/IO/DirectoryHelperTests.cs#L177)):*
<details><summary><i>Previously, ermshiperete (Eberhard Beilharz) wrote…</i></summary><blockquote>

Assert.Ignore("Can't test this on this system");
</blockquote></details>

Done.

---

*[SIL.Core.Tests/IO/DirectoryHelperTests.cs, line 196 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/622#-L6MTJ33YRPnZMT3Rg0d:-L6oOddCZ-S08Ywp1x59:b-896fix) ([raw file](https://github.com/sillsdev/libpalaso/blob/9eac80d73b6633abd88543988053b9e72221ad93/SIL.Core.Tests/IO/DirectoryHelperTests.cs#L196)):*
<details><summary><i>Previously, ermshiperete (Eberhard Beilharz) wrote…</i></summary><blockquote>

Looks like we should break out of the loop if we successfully set the current dir
</blockquote></details>

Done.

---

*[SIL.Core.Tests/IO/FileHelperTests.cs, line 1 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/622#-L6KXUrQMvHvvkTWD8nJ-r1-1:-L6oOeobKtg-y1kx8JWX:b-896fix) ([raw file](https://github.com/sillsdev/libpalaso/blob/9eac80d73b6633abd88543988053b9e72221ad93/SIL.Core.Tests/IO/FileHelperTests.cs#L1)):*
<details><summary><i>Previously, jasonleenaylor (Jason Naylor) wrote…</i></summary><blockquote>

header
</blockquote></details>

Done.

---

*[SIL.Core.Tests/IO/FileTestEnvironment.cs, line 1 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/622#-L6KXUrQMvHvvkTWD8nK-r1-1:-L6oOfypxysgxlR5r8Nk:b-896fix) ([raw file](https://github.com/sillsdev/libpalaso/blob/9eac80d73b6633abd88543988053b9e72221ad93/SIL.Core.Tests/IO/FileTestEnvironment.cs#L1)):*
<details><summary><i>Previously, jasonleenaylor (Jason Naylor) wrote…</i></summary><blockquote>

header
</blockquote></details>

Done.

---

*[SIL.Core.Tests/IO/RobustFileTests.cs, line 3 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/622#-L6KXUrPmNz3BcxL53Il-r1-3:-L6oOge6cbZ0A0wVMLb5:b-896fix) ([raw file](https://github.com/sillsdev/libpalaso/blob/9eac80d73b6633abd88543988053b9e72221ad93/SIL.Core.Tests/IO/RobustFileTests.cs#L3)):*
<details><summary><i>Previously, jasonleenaylor (Jason Naylor) wrote…</i></summary><blockquote>

more re-ordering
</blockquote></details>

Done.

---

*[SIL.Core/CommandLineProcessing/CommandLineRunner.cs, line 1 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/622#-L6KXUrPmNz3BcxL53Ix-r1-1:-L6oOhGJZhB04itfA-3_:b-896fix) ([raw file](https://github.com/sillsdev/libpalaso/blob/9eac80d73b6633abd88543988053b9e72221ad93/SIL.Core/CommandLineProcessing/CommandLineRunner.cs#L1)):*
<details><summary><i>Previously, jasonleenaylor (Jason Naylor) wrote…</i></summary><blockquote>

header
</blockquote></details>

Done.

---

*[SIL.Core/IO/DirectoryHelper.cs, line 1 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/622#-L6KXUrRuggFl7qImWko-r1-1:-L6oOhpXM5Fon78fs5lG:b-896fix) ([raw file](https://github.com/sillsdev/libpalaso/blob/9eac80d73b6633abd88543988053b9e72221ad93/SIL.Core/IO/DirectoryHelper.cs#L1)):*
<details><summary><i>Previously, jasonleenaylor (Jason Naylor) wrote…</i></summary><blockquote>

header
</blockquote></details>

Done.

---

*[SIL.Core/IO/DirectoryHelper.cs, line 76 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/622#-L6N2MMxd9VvGFZoRw-V:-L6oTDxb350xtQ9GpjBp:b-896fix) ([raw file](https://github.com/sillsdev/libpalaso/blob/9eac80d73b6633abd88543988053b9e72221ad93/SIL.Core/IO/DirectoryHelper.cs#L76)):*
<details><summary><i>Previously, ermshiperete (Eberhard Beilharz) wrote…</i></summary><blockquote>

I wonder if this works on  Linux if one of the two arguments ends in a forward slash. Do we have a test for that?
</blockquote></details>

Done.

---

*[SIL.Core/IO/FileHelper.cs, line 1 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/622#-L6KXUrRuggFl7qImWkp-r1-1:-L6oOjf2xrYt5lpMXOGE:b-896fix) ([raw file](https://github.com/sillsdev/libpalaso/blob/9eac80d73b6633abd88543988053b9e72221ad93/SIL.Core/IO/FileHelper.cs#L1)):*
<details><summary><i>Previously, jasonleenaylor (Jason Naylor) wrote…</i></summary><blockquote>

header
</blockquote></details>

Done.

---

*[SIL.Core/IO/FileHelper.cs, line 37 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/622#-L6N31_L5M8qwkIBJEtx:-L6oOloo_DDmPLIMm8Wg:b-896fix) ([raw file](https://github.com/sillsdev/libpalaso/blob/9eac80d73b6633abd88543988053b9e72221ad93/SIL.Core/IO/FileHelper.cs#L37)):*
<details><summary><i>Previously, ermshiperete (Eberhard Beilharz) wrote…</i></summary><blockquote>

If we find a match we don't call Close(), but I guess it's implicit because of the surrounding `using`. I suggest removing this line.
</blockquote></details>

Done.

---

*[SIL.Core/IO/PathHelper.cs, line 1 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/622#-L6KXUrRuggFl7qImWkq-r1-1:-L6oOmTXtU9aF6t2n31n:b-896fix) ([raw file](https://github.com/sillsdev/libpalaso/blob/9eac80d73b6633abd88543988053b9e72221ad93/SIL.Core/IO/PathHelper.cs#L1)):*
<details><summary><i>Previously, jasonleenaylor (Jason Naylor) wrote…</i></summary><blockquote>

header
</blockquote></details>

Done.

---

*[SIL.Core/IO/RobustIO.cs, line 1 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/622#-L6KXUrPmNz3BcxL53J6-r1-1:-L6oOn7CplClFYYOMPgv:b-896fix) ([raw file](https://github.com/sillsdev/libpalaso/blob/9eac80d73b6633abd88543988053b9e72221ad93/SIL.Core/IO/RobustIO.cs#L1)):*
<details><summary><i>Previously, jasonleenaylor (Jason Naylor) wrote…</i></summary><blockquote>

header
</blockquote></details>

Done.

---

*[SIL.Core/IO/TempFile.cs, line 1 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/622#-L6KXUrPmNz3BcxL53J7-r1-1:-L6oOneSFU2ZC0Ic2zW1:b-896fix) ([raw file](https://github.com/sillsdev/libpalaso/blob/9eac80d73b6633abd88543988053b9e72221ad93/SIL.Core/IO/TempFile.cs#L1)):*
<details><summary><i>Previously, jasonleenaylor (Jason Naylor) wrote…</i></summary><blockquote>

header
</blockquote></details>

Done.

---

*[SIL.Core/IO/TempFile.cs, line 5 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/622#-L6KXUrPmNz3BcxL53J7-r1-5:-L6oOo4BC2pPBoJok2eP:b-896fix) ([raw file](https://github.com/sillsdev/libpalaso/blob/9eac80d73b6633abd88543988053b9e72221ad93/SIL.Core/IO/TempFile.cs#L5)):*
<details><summary><i>Previously, jasonleenaylor (Jason Naylor) wrote…</i></summary><blockquote>

remove
</blockquote></details>

Done.

---

*[SIL.Core/PlatformUtilities/Platform.cs, line 14 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/622#-L6N5s8pNu8jGsSSXv4J:-L6oTJgcu8DqI6qEIG11:bf1aob2) ([raw file](https://github.com/sillsdev/libpalaso/blob/9eac80d73b6633abd88543988053b9e72221ad93/SIL.Core/PlatformUtilities/Platform.cs#L14)):*
<details><summary><i>Previously, ermshiperete (Eberhard Beilharz) wrote…</i></summary><blockquote>

@chrisvire Since I think you wrote the original code that made the distinction  between Linux and Mac, maybe you want to take a look at this change if this works as expected on Mac.
</blockquote></details>

This is the new and recommended way to determine the platform. It is part of the BCL for .NET Core, .NET Standard, and .NET Framework 4.7.1+.

---

*[SIL.Windows.Forms.Tests/SettingsProviderTests.cs, line 154 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/622#-L6MbEIoikMUYuPTZueZ:-L6oOqBprxfBf7zqpZsf:b-896fix) ([raw file](https://github.com/sillsdev/libpalaso/blob/9eac80d73b6633abd88543988053b9e72221ad93/SIL.Windows.Forms.Tests/SettingsProviderTests.cs#L154)):*
<details><summary><i>Previously, ermshiperete (Eberhard Beilharz) wrote…</i></summary><blockquote>

Can you add a comment what the problem is? IRC it's the buggy settings implementation that we fixed in our mono*-sil, but that hasn't made it in an upstream release yet.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/sillsdev/libpalaso/622)*
<!-- Sent from Reviewable.io -->

ermshiperete commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 188 files at r1, 27 of 27 files at r2.
Review status: all files reviewed at latest revision, 16 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/sillsdev/libpalaso/622#-:-L6pF6tfr2gy5saw7FDZ:bnfp4nl)*
<!-- Sent from Reviewable.io -->

ddaspit commented:
Review status: 146 of 154 files reviewed at latest revision, 16 unresolved discussions, some commit checks failed.

---

*[SIL.BuildTasks/StampAssemblies/StampAssemblies.cs, line 193 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/622#-L6KXUrPmNz3BcxL53Id-r1-193:-L6tUbGd1uQvBqSj6S6C:b1w2ktm) ([raw file](https://github.com/sillsdev/libpalaso/blob/9eac80d73b6633abd88543988053b9e72221ad93/SIL.BuildTasks/StampAssemblies/StampAssemblies.cs#L193)):*
<details><summary><i>Previously, ddaspit (Damien Daspit) wrote…</i></summary><blockquote>

I think this API was only public for unit test purposes. If projects are using it, I can keep the old API as well.
</blockquote></details>

I updated the class so that it maintains the current API just in case someone uses it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/sillsdev/libpalaso/622)*
<!-- Sent from Reviewable.io -->

ddaspit commented:
I put the code back that deletes the temporary files when XslCompiledTransform is used. It is conditionally compiled and will only be included in the .NET Framework build of SIL.Core.

ermshiperete commented:
Reviewed 10 of 10 files at r3.
Review status: all files reviewed at latest revision, 17 unresolved discussions, some commit checks failed.

---

*[SIL.BuildTasks/StampAssemblies/StampAssemblies.cs, line 22 at r3](https://reviewable.io:443/reviews/sillsdev/libpalaso/622#-L6ubCZiylQ_pkgiC3HQ:-L6ubCZjrKvcnzmgRPJh:b-ogmryi) ([raw file](https://github.com/sillsdev/libpalaso/blob/248371c59d230121329cffa5b50c83a1e21463e5/SIL.BuildTasks/StampAssemblies/StampAssemblies.cs#L22)):*
> ```Smalltalk
> 		{
> 			public string[] parts = new string[4];
> 			public string prerelease;
> ```

Can we make this private or the name uppercase since that gets newly introduced?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/sillsdev/libpalaso/622)*
<!-- Sent from Reviewable.io -->

ddaspit commented:
Review status: all files reviewed at latest revision, 17 unresolved discussions, some commit checks failed.

---

*[SIL.BuildTasks/StampAssemblies/StampAssemblies.cs, line 22 at r3](https://reviewable.io:443/reviews/sillsdev/libpalaso/622#-L6ubCZiylQ_pkgiC3HQ:-L6uvNqY-l3BnFcwUjrS:b-896fix) ([raw file](https://github.com/sillsdev/libpalaso/blob/248371c59d230121329cffa5b50c83a1e21463e5/SIL.BuildTasks/StampAssemblies/StampAssemblies.cs#L22)):*
<details><summary><i>Previously, ermshiperete (Eberhard Beilharz) wrote…</i></summary><blockquote>

Can we make this private or the name uppercase since that gets newly introduced?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/sillsdev/libpalaso/622)*
<!-- Sent from Reviewable.io -->

ermshiperete commented:
Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 16 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/sillsdev/libpalaso/622)*
<!-- Sent from Reviewable.io -->

jasonleenaylor commented:
This PR review makes me miss gerrit. You have my approval if @ermshiperete is satisfied.

ermshiperete commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 16 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/sillsdev/libpalaso/622#-:-L6zF_jLf4tF8aamPVT9:bnfp4nl)*
<!-- Sent from Reviewable.io -->

ddaspit commented:
Reviewed 1 of 188 files at r1, 5 of 5 files at r6.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/sillsdev/libpalaso/622)*
<!-- Sent from Reviewable.io -->

ddaspit commented:
I temporarily disabled the Linux build status check to merge this, because build configuration changes were required to build correctly. This was checked using a netstandard build configuration, which built successfully.
